benches/sync_notify.rs

```diff
@@ -58,18 +61,20 @@ fn notify_one<const N_WAITERS: usize>(b: &mut Bencher) {
     }
 
     const N_ITERS: usize = 500;
-    b.iter(|| {
-        counter.store(0, Ordering::Relaxed);
-        loop {
-            notify.notify_one();
-            if counter.load(Ordering::Relaxed) >= N_ITERS {
-                break;
+    c.bench_function("notify_one", |b| {
+        b.iter(|| {
+            counter.store(0, Ordering::Relaxed);
+            loop {
+                notify.notify_one();
+                if counter.load(Ordering::Relaxed) >= N_ITERS {
+                    break;
+                }
             }
-        }
+        })
     });
 }
 
-bencher::benchmark_group!(
+criterion_group!(
     notify_waiters_simple,
```

Comment:
With `criterion` crate, we can group the similar benchmarks in one group.
Let me use the `benches/sync_notify.rs` as example to expalin this.

The `notify_one::<10>`、`notify_one::<100>`..., they almost same, but only the `N_WAITERS` is not equal.

So, we can group these benchmarks:

```rust
use std::sync::Arc;
use std::sync::atomic::{AtomicUsize, Ordering};

use criterion::{BenchmarkGroup, Criterion, criterion_group, criterion_main};
use criterion::measurement::WallTime;

use tokio::sync::Notify;

fn rt() -> tokio::runtime::Runtime {
    tokio::runtime::Builder::new_multi_thread()
        .worker_threads(6)
        .build()
        .unwrap()
}

fn notify_waiters<const N_WAITERS: usize>(group: &mut BenchmarkGroup<WallTime>) {
    let rt = rt();
    let notify = Arc::new(Notify::new());
    let counter = Arc::new(AtomicUsize::new(0));

    for _ in 0..N_WAITERS {
        rt.spawn({
            let notify = notify.clone();
            let counter = counter.clone();
            async move {
                loop {
                    notify.notified().await;
                    counter.fetch_add(1, Ordering::Relaxed);
                }
            }
        });
    }

    const N_ITERS: usize = 500;
    group.bench_function(N_WAITERS.to_string(), |b| b.iter(||
        {
            counter.store(0, Ordering::Relaxed);
            loop {
                notify.notify_waiters();
                if counter.load(Ordering::Relaxed) >= N_ITERS {
                    break;
                }
            }
        }
    ));
}

fn notify_one<const N_WAITERS: usize>(group: &mut BenchmarkGroup<WallTime>) {
    let rt = rt();
    let notify = Arc::new(Notify::new());
    let counter = Arc::new(AtomicUsize::new(0));
    for _ in 0..N_WAITERS {
        rt.spawn({
            let notify = notify.clone();
            let counter = counter.clone();
            async move {
                loop {
                    notify.notified().await;
                    counter.fetch_add(1, Ordering::Relaxed);
                }
            }
        });
    }

    const N_ITERS: usize = 500;
    group.bench_function(N_WAITERS.to_string(), |b| b.iter(||
        {
            counter.store(0, Ordering::Relaxed);
            loop {
                notify.notify_one();
                if counter.load(Ordering::Relaxed) >= N_ITERS {
                    break;
                }
            }
        }
    ));
}

fn bench_notify_waiters(c: &mut Criterion) {
    let mut group = c.benchmark_group("notify_waiters");
    notify_waiters::<10>(&mut group);
    notify_waiters::<50>(&mut group);
    notify_waiters::<100>(&mut group);
    notify_waiters::<200>(&mut group);
    notify_waiters::<500>(&mut group);
    group.finish();
}

fn bench_notify_one(c: &mut Criterion) {
    let mut group = c.benchmark_group("notify_one");
    notify_one::<10>(&mut group);
    notify_one::<50>(&mut group);
    notify_one::<100>(&mut group);
    notify_one::<200>(&mut group);
    notify_one::<500>(&mut group);
    group.finish();
}

criterion_group!(
    notify_waiters_simple,
    bench_notify_waiters,
    bench_notify_one,
);

criterion_main!(notify_waiters_simple);

```

The benchmark output is the following:

```rust
notify_waiters/10       time:   [136.88 µs 142.21 µs 148.36 µs]
                        change: [+4.8894% +7.8281% +11.130%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 16 outliers among 100 measurements (16.00%)
  3 (3.00%) high mild
  13 (13.00%) high severe
notify_waiters/50       time:   [128.05 µs 129.35 µs 130.75 µs]
                        change: [+1.4689% +3.0748% +4.6824%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe
notify_waiters/100      time:   [119.49 µs 120.84 µs 122.25 µs]
                        change: [-3.9765% -2.1273% +0.0038%] (p = 0.04 < 0.05)
                        Change within noise threshold.
....
```

Does it is more readable than current version output:
```rust
notify_waiters          time:   [129.83 µs 131.15 µs 132.53 µs]
                        change: [-7.7179% -4.5238% -1.1667%] (p = 0.01 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe

notify_waiters #2       time:   [131.63 µs 133.83 µs 136.14 µs]
                        change: [-3.7356% -1.2708% +0.8673%] (p = 0.31 > 0.05)
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

notify_waiters #3       time:   [130.58 µs 134.51 µs 139.47 µs]
                        change: [+16.498% +26.125% +36.543%] (p = 0.00 < 0.05)
                        Performance has regressed.
....
```

Reply:
I agree. is there anywhere else in the benchmarks you would like to see them grouped?

Reply:
I belive the benchmarks of components in `sync` mod can be grouped, including the `sync_mpsc.rs`, `sync_rwlock.rs`, `sync_semaphore.rs`.  They might can be grouped by `contention or uncontention`, `new_multi_thread or new_current_thread`, and `full or unfull`. But I dont want them to be grouped too complicatedly, grouping them once might be enough.

And could we use generics to simplify the benchmark like `sync_notify.rs` does?

Reply:
I tried to use generics where it made sense and grouped benchmarks together by `contented or uncontented`.